File: src/chunking/heading_chunker.py
from collections import Counter
from src.chunking.base import Chunker
from src.chunking.token_window_chunker import TokenWindowChunker
from src.models import DocumentPage, Section, TextLine
from src.tokenizer import TiktokenCounter
HEADING_SIZE_RATIO = 1.15
MAX_HEADING_WORDS = 14
SENTENCE_ENDINGS = (".", ",", ";")
BRACKET_PAIRS = (("(", ")"), ("[", "]"), ("{", "}"))
DEFAULT_MIN_TOKENS = 20
DEFAULT_MAX_TOKENS = 400
def _body_font_size(lines: list[TextLine]) -> float | None:
    weights: Counter[float] = Counter()
    for line in lines:
        if line.font_size is not None and not line.is_bold:
            weights[line.font_size] += len(line.text)
    if not weights:
        return None
    return weights.most_common(1)[0][0]

def _is_heading(line: TextLine, body_size: float | None) -> bool:
    if line.heading_level is not None:
        return True
    if body_size is None or line.font_size is None:
        return False
    if len(line.text.split()) > MAX_HEADING_WORDS:
        return False
    if line.text.endswith(SENTENCE_ENDINGS):
        return False
    return line.is_bold or line.font_size > body_size * HEADING_SIZE_RATIO

def _is_unterminated(text: str) -> bool:
    if text.count('"') % 2 == 1:
        return True
    return any(text.count(o) > text.count(c) for o, c in BRACKET_PAIRS)

class HeadingChunker(Chunker):
    def __init__(
        self,
        min_tokens: int = DEFAULT_MIN_TOKENS,
        max_tokens: int = DEFAULT_MAX_TOKENS,
        counter: TiktokenCounter | None = None,
    ):
        self.min_tokens = min_tokens
        self.max_tokens = max_tokens
        self.counter = counter or TiktokenCounter()

    def split(self, pages: list[DocumentPage]) -> list[Section]:
        lines = [line for page in pages for line in page.lines]
        if not lines:
            return []
        body_size = _body_font_size(lines)
        raw: list[dict] = []
        current: dict | None = None
        for line in lines:
            if _is_heading(line, body_size):
                if (
                    current is not None
                    and not current["body"]
                    and _is_unterminated(current["title"])
                ):
                    current["title"] = f"{current['title']} {line.text}"
                    continue
                current = {
                    "title": line.text,
                    "page_number": line.page_number,
                    "body": [],
                }
                raw.append(current)
            elif current is not None:
                current["body"].append(line.text)
        return self._build_sections(raw, pages[0].source)

    def _build_sections(self, raw: list[dict], source: str) -> list[Section]:
        parent: str | None = None
        sections: list[Section] = []
        windower = TokenWindowChunker(
            max_tokens = self.max_tokens, counter = self.counter
        )
        for item in raw:
            body = "\n".join(item["body"]).strip()
            if not body:
                parent = item["title"]
                continue
            tokens = self.counter.count(body)
            if sections and tokens < self.min_tokens:
                previous = sections[-1]
                previous.text = f"{previous.text}\n{item['title']}\n{body}"
                previous.token_count = self.counter.count(previous.text)
                continue
            pieces = windower.windows(body) if tokens > self.max_tokens else [body]
            for index, piece in enumerate(pieces, start = 1):
                title = item["title"]
                if len(pieces) > 1:
                    title = f"{title} (part {index}/{len(pieces)})"
                sections.append(
                    Section(
                        section_id = f"s{len(sections) + 1:02d}",
                        title = title,
                        text = piece,
                        source = source,
                        page_number = item["page_number"],
                        parent_heading = parent,
                        token_count = self.counter.count(piece),
                    )
                )
        return sections

File: src/chunking/headingChunker.ts
import { TiktokenCounter } from "../tokenizer.ts";
import { TokenWindowChunker } from "./tokenWindowChunker.ts";
import type { Chunker } from "./base.ts";
import type { DocumentPage, Section, TextLine } from "../models.ts";
import type { TokenCounter } from "../tokenizer.ts";

const HEADING_SIZE_RATIO = 1.15; // 15% bigger than body text -> probably a heading
const MAX_HEADING_WORDS = 14;
const SENTENCE_ENDINGS = [".", ",", ";"];
const BRACKET_PAIRS: readonly [string, string][] = [
  ["(", ")"],
  ["[", "]"],
  ["{", "}"],
];
const DEFAULT_MIN_TOKENS = 20;
const DEFAULT_MAX_TOKENS = 400;

/** The most-used non-bold font size, weighted by how much text is set in it. */
function bodyFontSize(lines: TextLine[]): number | null {
  const weights = new Map<number, number>();
  for (const line of lines) {
    if (line.fontSize == null || line.isBold) continue;
    weights.set(line.fontSize, (weights.get(line.fontSize) ?? 0) + line.text.length);
  }
  if (weights.size === 0) return null;
  return [...weights].reduce((a, b) => (b[1] > a[1] ? b : a))[0];
}

function isHeading(line: TextLine, bodySize: number | null): boolean {
  // Markdown and Word tell us outright; PDFs make us guess from the typography.
  if (line.headingLevel != null) return true;
  if (bodySize == null || line.fontSize == null) return false;
  if (line.text.split(/\s+/).length > MAX_HEADING_WORDS) return false;
  if (SENTENCE_ENDINGS.some((end) => line.text.endsWith(end))) return false;
  return Boolean(line.isBold) || line.fontSize > bodySize * HEADING_SIZE_RATIO;
}

function count(text: string, needle: string): number {
  return text.split(needle).length - 1;
}

/** A heading that hangs open — an unclosed quote or bracket — is a wrapped heading. */
function isUnterminated(text: string): boolean {
  if (count(text, '"') % 2 === 1) return true;
  return BRACKET_PAIRS.some(([open, close]) => count(text, open) > count(text, close));
}

interface RawSection {
  title: string;
  pageNumber: number;
  body: string[];
}

export interface HeadingChunkerOptions {
  minTokens?: number;
  maxTokens?: number;
  counter?: TokenCounter;
}

/** The preferred splitter: cut the document where its own headings say to. */
export class HeadingChunker implements Chunker {
  readonly minTokens: number;
  readonly maxTokens: number;
  readonly counter: TokenCounter;

  constructor(options: HeadingChunkerOptions = {}) {
    this.minTokens = options.minTokens ?? DEFAULT_MIN_TOKENS;
    this.maxTokens = options.maxTokens ?? DEFAULT_MAX_TOKENS;
    this.counter = options.counter ?? new TiktokenCounter();
  }

  split(pages: DocumentPage[]): Section[] {
    const lines = pages.flatMap((page) => page.lines);
    const first = pages[0];
    if (lines.length === 0 || !first) return [];

    const bodySize = bodyFontSize(lines);
    const raw: RawSection[] = [];
    let current: RawSection | null = null;

    for (const line of lines) {
      if (isHeading(line, bodySize)) {
        // two heading lines in a row, the first left hanging -> one wrapped heading
        if (current && current.body.length === 0 && isUnterminated(current.title)) {
          current.title = `${current.title} ${line.text}`;
          continue;
        }
        current = { title: line.text, pageNumber: line.pageNumber, body: [] };
        raw.push(current);
      } else if (current) {
        current.body.push(line.text);
      }
    }
    return this.buildSections(raw, first.source);
  }

  private buildSections(raw: RawSection[], source: string): Section[] {
    const sections: Section[] = [];
    const windower = new TokenWindowChunker({
      maxTokens: this.maxTokens,
      counter: this.counter,
    });
    let parent: string | null = null;

    for (const item of raw) {
      const body = item.body.join("\n").trim();
      if (!body) {
        // a heading with nothing under it is a parent heading for what follows
        parent = item.title;
        continue;
      }

      const tokens = this.counter.count(body);
      const previous = sections.at(-1);
      if (previous && tokens < this.minTokens) {
        // too small to stand alone: fold it into the section before it
        previous.text = `${previous.text}\n${item.title}\n${body}`;
        previous.tokenCount = this.counter.count(previous.text);
        continue;
      }

      const pieces = tokens > this.maxTokens ? windower.windows(body) : [body];
      for (const [index, piece] of pieces.entries()) {
        const title =
          pieces.length > 1
            ? `${item.title} (part ${index + 1}/${pieces.length})`
            : item.title;
        sections.push({
          sectionId: `s${String(sections.length + 1).padStart(2, "0")}`,
          title,
          text: piece,
          source,
          pageNumber: item.pageNumber,
          parentHeading: parent,
          tokenCount: this.counter.count(piece),
        });
      }
    }
    return sections;
  }
}
